code/Style.md

```diff
@@ -70,6 +72,38 @@ func Example() (err error) {
 }
 ```
 
+There are __2 exceptions__ when an error **NEVER MUST BE** wrapped nor annotated nor combined:
+
+1. Those functions which return the [`io.EOF`](https://golang.org/pkg/io/#pkg-variables) to indicate that they have finished and they don't have to be called again, like the types which implement the [`io.Reader` interface](https://golang.org/pkg/io/#Reader).
+2. Some packages which require that user specific functions return specific error types for acting appropriately, like [_gRPC status](https://godoc.org/google.golang.org/grpc/status) for being able to respond a [specific protocol error status code](https://github.com/grpc/grpc/blob/master/doc/statuscodes.md).
+
```

Comment:
There are more exceptions, unfortunately. Anywhere 3rd party code relies on a particular value in an error, we can't wrap it without breaking things. For example, I recently ran into code that checks `os.IsNotExist` and `os.IsExist` on error values. Those don't work anymore when matching errors are wrapped.

How does one know for sure whether such values might be in an error? In general, we can't, but I suppose we could try to build up a big list of things to check for. I would much, much rather have us come up with something simpler for the developer, though. Maybe an automated checker that notices when a type implements `io.Reader` or `io.Writer` (etc) and complains when errors are wrapped unconditionally? Maybe we use a special different type from `error` when we expect wrapped things? Maybe we use some kind of special annotation on functions whose error return values should not be wrapped unconditionally?

In short, this seems like a big flaw with our current approach.

Reply:
I can change the `1` to be more generic; it isn't very shiny as a full list or your ideas, but I think that it's the first step to improve our docs and doing it incrementally, otherwise, we wouldn't have any until we can dedicate resources and time to work on those.

Let me know if that's fine, otherwise, I will close this PR without merging it.

Reply:
Even if we want to automate these checks, we need to first list out all the situations where we need to preserve the errors.

Reply:
hence, should we start with this one and more through the time that we realize?

Reply:
Yeah, we can start automating later.

Reply:
@thepaul I modified, please read again if that's better for you as the first starting point.